File: tools/dogfood_coverage.py
#!/usr/bin/env python3
"""Dogfood-coverage scorecard: how much of the REAL dev workflow eats the kernel.

The fak thesis is that the kernel (`fak serve` / `fak guard`) belongs in front of
every tool call an agent makes. "Dogfooding" only counts when *our own* dev work —
the dispatch fleet, the interactive Claude Code session — actually crosses that
boundary. This scorecard measures it, and (like every other fak scorecard) it
cross-checks REALITY, not config: it imports `dispatch_worker` and calls the live
`guarded_launch_command` on THIS host, and it counts decision rows in the durable
audit journals — so the number cannot be gamed by editing a flag.

KPIs fold into one `coverage` percent + a `dogfood_debt` integer (count of unmet
HARD affordances) + an A–F grade, and emit a control-pane JSON payload. Drive it on
a /loop cadence to keep the dev loop kernel-adjudicated.

    python tools/dogfood_coverage.py            # human report
    python tools/dogfood_coverage.py --json      # control-pane payload
    python tools/dogfood_coverage.py --check      # exit 1 if any HARD KPI is unmet
"""
from __future__ import annotations

import argparse
import importlib.util
import json
import os
import sys
from pathlib import Path
from typing import Any


def repo_root(start: Path | None = None) -> Path:
    here = (start or Path(__file__)).resolve()
    return here.parent.parent


def _load_module(name: str, path: Path) -> Any:
    """Import a sibling tool by path (no package install needed)."""
    spec = importlib.util.spec_from_file_location(name, path)
    if not (spec and spec.loader):
        raise ImportError(f"cannot load {name} from {path}")
    mod = importlib.util.module_from_spec(spec)
    sys.path.insert(0, str(path.parent))
    spec.loader.exec_module(mod)
    return mod


def _grep(path: Path, needle: str) -> bool:
    try:
        return needle in path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return False


def count_audit_rows(root: Path) -> tuple[int, int]:
    """Count decision rows the kernel recorded across the fleet's per-lane decision
    journals (the WITNESS that real dogfood happened, not just got configured).

    Returns (rows, journals). Each non-empty JSONL line is one kernel verdict. The
    per-lane journals live under the gitignored .dispatch-runs/guard-audit/; the
    interactive `fak guard` default writes one under the user config dir, which we
    add when present.
    """
    rows = 0
    journals = 0
    candidates: list[Path] = []
    fleet_dir = root / ".dispatch-runs" / "guard-audit"
    if fleet_dir.is_dir():
        candidates.extend(sorted(fleet_dir.glob("*.jsonl")))
    # The interactive front door's per-user default journal is host-global, so
    # include it only for real repo roots. Temp fixture roots should count only
    # the journals they create, otherwise a developer's live audit history makes
    # hermetic tests nondeterministic.
    if (root / ".git").exists():
        cfg = os.environ.get("XDG_CONFIG_HOME") or os.environ.get("APPDATA")
        if not cfg:
            home = os.environ.get("HOME") or os.path.expanduser("~")
            cfg = str(Path(home) / ".config")
        user_journal = Path(cfg) / "fak" / "guard-audit.jsonl"
        if user_journal.is_file():
            candidates.append(user_journal)
    for jp in candidates:
        try:
            text = jp.read_text(encoding="utf-8", errors="replace")
        except OSError:
            continue
        n = sum(1 for ln in text.splitlines() if ln.strip())
        if n:
            rows += n
            journals += 1
    return rows, journals


def diagnose_audit_gap(root: Path) -> str:
    """Explain WHY the audit journal is empty, so a zero row count is self-diagnosing
    instead of an undifferentiated 0 (the default-observability point: a blank witness
    must say which blank it is). Returns "" when rows exist (no gap to explain).

    The three blanks a `0` can hide are distinct actions:
      * no journal DIR at all      -> no guarded worker has ever run here
      * journal dir, only empty/blank files -> guard booted but the wrapped agent
        proposed no tool call the kernel adjudicated (e.g. it exited before its first
        tool use — the silent auth-failure / empty-turn case)
      * dir present, files present, rows>0 -> not a gap (handled by the caller)
    """
    fleet_dir = root / ".dispatch-runs" / "guard-audit"
    if not fleet_dir.is_dir():
        return ("no guard-audit journal directory yet — no guarded worker has run on "
                "this host (arm `fak guard -- <agent>` so the kernel records verdicts)")
    jsonls = sorted(fleet_dir.glob("*.jsonl"))
    if not jsonls:
        return ("guard-audit directory exists but holds no journal files — the guard "
                "wire is configured but never exercised by a launched worker")
    # Dir + files present but every file is blank: the worker booted under guard but
    # never reached a tool call the kernel adjudicated (the silent empty-turn signature
    # — e.g. the wrapped agent exited on an auth/login failure before its first tool use).
    return (f"{len(jsonls)} journal file(s) present but all blank — a guarded worker "
            "booted but proposed no adjudicated tool call (check the agent reached a "
            "tool use; an auth/login failure exits before the first verdict)")


def evaluate(root: Path, env: dict[str, str] | None = None) -> dict[str, Any]:
    """Compute the dogfood-coverage KPIs. `env` defaults to the process env so the
    LIVE opt-out (FLEET_DOGFOOD_GUARD=0) shows up in the score."""
    e = env if env is not None else dict(os.environ)

    kpis: list[dict[str, Any]] = []

    def add(key: str, ok: bool, hard: bool, detail: str, evidence: str = "") -> None:
        kpis.append({"key": key, "ok": bool(ok), "hard": hard,
                     "detail": detail, "evidence": evidence})

    # KPIs 1-3 cross-check the LIVE dispatch_worker surface (a real call, not a grep).
    # If that module moves or a symbol it calls gets renamed, this must degrade the
    # three dependent KPIs to an honest unmet-with-reason -- not raise ImportError/
    # AttributeError/OSError and crash the whole scorecard (#1941).
    try:
        dw = _load_module("dispatch_worker", root / "tools" / "dispatch_worker.py")
        raw = dw.build_command("probe", "claude")
        launch_cmd, guarded = dw.guarded_launch_command(raw, "probe", "claude", root, env=e)
        fak_bin = dw.resolve_fak_bin(root, e)
        live_on = dw.guard_enabled(e)
    except (ImportError, AttributeError, OSError) as exc:
        reason = f"dispatch surface moved: {exc}"
        add("fleet_leaf_guarded", False, True,
            "dispatch_worker.guarded_launch_command fronts a claude worker with `fak guard`",
            evidence=reason)
        add("fak_bin_resolvable", False, True,
            "a `fak` binary resolves (FAK_BIN / tools/.bin / PATH) so guard-wrapping engages",
            evidence=reason)
        add("guard_default_on", False, True,
            "FLEET_DOGFOOD_GUARD is not disabled in the live environment (default ON)",
            evidence=reason)
    else:
        # 1. The supervised leaf launcher fronts a claude worker with the kernel, on
        #    THIS host, RIGHT NOW. This is the real behavior cross-check — not a grep.
        add("fleet_leaf_guarded", guarded, True,
            "dispatch_worker.guarded_launch_command fronts a claude worker with `fak guard`",
            evidence=" ".join(launch_cmd[:3]) if guarded else "UNWRAPPED (coverage=0 on this host)")

        # 2. A `fak` binary resolves, so the fail-open path is not silently dropping the
        #    fleet to 0% coverage.
        add("fak_bin_resolvable", bool(fak_bin), True,
            "a `fak` binary resolves (FAK_BIN / tools/.bin / PATH) so guard-wrapping engages",
            evidence=fak_bin or "no fak binary found — workers run UNGUARDED")

        # 3. Dogfood mode is ON in the live env (default-on, not opted out here).
        add("guard_default_on", live_on, True,
            "FLEET_DOGFOOD_GUARD is not disabled in the live environment (default ON)",
            evidence=f"FLEET_DOGFOOD_GUARD={e.get('FLEET_DOGFOOD_GUARD', '<unset=ON>')}")

    # 4. The scheduled-task lane (issue_dispatch) is wired to the same guard path.
    wired = _grep(root / "tools" / "issue_dispatch.py", "guarded_launch_command")
    add("issue_dispatch_wired", wired, True,
        "issue_dispatch.evaluate routes its detached spawn through guarded_launch_command",
        evidence="tools/issue_dispatch.py calls guarded_launch_command" if wired else "MISSING")

    # 5. The interactive front door exists as a productized verb.
    guard_go = (root / "cmd" / "fak" / "guard.go").is_file()
    add("guard_verb_present", guard_go, True,
        "`fak guard -- claude` exists (cmd/fak/guard.go) as the one-command kernel front door",
        evidence="cmd/fak/guard.go" if guard_go else "MISSING")

    # 6. The front door is documented for a human/agent to find.
    documented = _grep(root / "DOGFOOD-CLAUDE.md", "fak guard")
    add("guard_documented", documented, False,
        "DOGFOOD-CLAUDE.md documents the `fak guard` front door", evidence="DOGFOOD-CLAUDE.md")

    # 7. Durable witness: kernel decisions actually recorded by guarded workers. Soft
    #    because a freshly-configured host has 0 until the fleet runs — but it is the
    #    proof the wire is exercised, not merely wired.
    rows, journals = count_audit_rows(root)
    audit_evidence = f"{rows} decision row(s) across {journals} journal(s)"
    if rows == 0:
        # Make the blank witness self-diagnosing: say WHICH zero this is so the
        # operator gets the unblock action, not an undifferentiated 0.
        audit_evidence += f" — {diagnose_audit_gap(root)}"
    add("audit_journal_evidence", rows > 0, False,
        "guarded workers have recorded kernel decisions in a durable audit journal",
        evidence=audit_evidence)

    # 8/9. Always-on compute so the dogfood loop runs 24/7 (the goal's other half).
    doc = (root / "docs" / "fak" / "always-on-dogfood-server.md").is_file()
    add("always_on_server_doc", doc, False,
        "an always-on dogfood server design exists (Mac/GCP tiers)",
        evidence="docs/fak/always-on-dogfood-server.md" if doc else "MISSING")
    plist = (root / "tools" / "com.fak.serve-gateway.plist").is_file()
    add("always_on_serve_plist", plist, False,
        "a launchd unit keeps a shared `fak serve` gateway alive 24/7",
        evidence="tools/com.fak.serve-gateway.plist" if plist else "MISSING")

    total = len(kpis)
    met = sum(1 for k in kpis if k["ok"])
    hard_unmet = [k for k in kpis if k["hard"] and not k["ok"]]
    coverage = round(100.0 * met / total, 1) if total else 0.0
    dogfood_debt = len(hard_unmet)
    grade = _grade(coverage, dogfood_debt)

    return {
        "schema": "dogfood-coverage/1",
        "coverage": coverage,
        "met": met,
        "total": total,
        "dogfood_debt": dogfood_debt,
        "grade": grade,
        "audit_rows": rows,
        "kpis": kpis,
        "worst_first": [k["key"] for k in hard_unmet] + [k["key"] for k in kpis if not k["hard"] and not k["ok"]],
    }


def _grade(coverage: float, debt: int) -> str:
    if debt == 0 and coverage >= 95:
        return "A"
    if debt == 0:
        return "B"
    if debt <= 1:
        return "C"
    if debt <= 2:
        return "D"
    return "F"


def render(payload: dict[str, Any]) -> str:
    lines = [
        f"dogfood-coverage: {payload['coverage']}% "
        f"({payload['met']}/{payload['total']} KPIs)  "
        f"grade {payload['grade']}  dogfood_debt {payload['dogfood_debt']}  "
        f"audit_rows {payload['audit_rows']}",
    ]
    for k in payload["kpis"]:
        mark = "OK " if k["ok"] else ("XX " if k["hard"] else ".. ")
        tag = "HARD" if k["hard"] else "soft"
        lines.append(f"  [{mark}] {k['key']:<24} ({tag})  {k['detail']}")
        if k.get("evidence"):
            lines.append(f"          -> {k['evidence']}")
    if payload["worst_first"]:
        lines.append("  next: " + ", ".join(payload["worst_first"]))
    return "\n".join(lines)


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description="Measure how much of the dev workflow eats the kernel.")
    ap.add_argument("--json", action="store_true", help="emit the control-pane JSON payload")
    ap.add_argument("--check", action="store_true", help="exit 1 if any HARD KPI is unmet")
    ap.add_argument("--workspace", default="", help="repo root (default: auto)")
    args = ap.parse_args(argv)

    root = Path(args.workspace).resolve() if args.workspace else repo_root()
    payload = evaluate(root)
    if args.json:
        print(json.dumps(payload, indent=2))
    else:
        print(render(payload))
    if args.check and payload["dogfood_debt"] > 0:
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/dogfoodcoverage/dogfoodcoverage.go
package dogfoodcoverage

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"time"
)

const (
	// Schema is the JSON schema version identifier for dogfood-coverage payloads.
	Schema = "dogfood-coverage/1"
)

// KPI represents a single evaluated key performance indicator.
type KPI struct {
	Key      string `json:"key"`
	OK       bool   `json:"ok"`
	Hard     bool   `json:"hard"`
	Detail   string `json:"detail"`
	Evidence string `json:"evidence"`
}

// Report captures the full dogfood-coverage scorecard payload.
type Report struct {
	Schema      string   `json:"schema"`
	Coverage    float64  `json:"coverage"`
	Met         int      `json:"met"`
	Total       int      `json:"total"`
	DogfoodDebt int      `json:"dogfood_debt"`
	Grade       string   `json:"grade"`
	AuditRows   int      `json:"audit_rows"`
	KPIs        []KPI    `json:"kpis"`
	WorstFirst  []string `json:"worst_first"`
}

var guardOffValues = map[string]struct{}{
	"0":       {},
	"off":     {},
	"false":   {},
	"no":      {},
	"disable": {},
}

func envLookup(env map[string]string, key string) (string, bool) {
	if env != nil {
		val, ok := env[key]
		return val, ok
	}
	return os.LookupEnv(key)
}

func envGet(env map[string]string, key string) string {
	val, _ := envLookup(env, key)
	return val
}

func isFile(path string) bool {
	fi, err := os.Stat(path)
	return err == nil && !fi.IsDir()
}

func fileContains(path, needle string) bool {
	data, err := os.ReadFile(path)
	if err != nil {
		return false
	}
	return strings.Contains(string(data), needle)
}

func findRepoRoot(start string) string {
	if start == "" {
		if wd, err := os.Getwd(); err == nil {
			start = wd
		}
	}
	abs, err := filepath.Abs(start)
	if err == nil {
		start = abs
	}
	dir := start
	for {
		if _, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return start
		}
		dir = parent
	}
}

func whichOnPath(name, pathVal string) string {
	if pathVal == "" {
		pathVal = os.Getenv("PATH")
	}
	suffixes := []string{""}
	if runtime.GOOS == "windows" && filepath.Ext(name) == "" {
		pathext := os.Getenv("PATHEXT")
		if pathext == "" {
			pathext = ".COM;.EXE;.BAT;.CMD"
		}
		for _, ext := range strings.Split(pathext, string(os.PathListSeparator)) {
			ext = strings.TrimSpace(ext)
			if ext != "" {
				suffixes = append(suffixes, strings.ToLower(ext), strings.ToUpper(ext))
			}
		}
	}

	dirs := strings.Split(pathVal, string(os.PathListSeparator))
	for _, dir := range dirs {
		if dir == "" {
			continue
		}
		for _, suffix := range suffixes {
			candidate := filepath.Join(dir, name+suffix)
			if fi, err := os.Stat(candidate); err == nil && !fi.IsDir() {
				return candidate
			}
		}
	}
	return ""
}

func resolveFakBin(root string, env map[string]string) string {
	explicit := strings.TrimSpace(envGet(env, "FAK_BIN"))
	if explicit != "" {
		if isFile(explicit) {
			return explicit
		}
	}

	var candidates []string
	pathVal := envGet(env, "PATH")
	onPath := whichOnPath("fak", pathVal)
	if onPath != "" {
		candidates = append(candidates, onPath)
	}

	exe := "fak"
	if runtime.GOOS == "windows" {
		exe = "fak.exe"
	}

	inTree := filepath.Join(root, "tools", ".bin", exe)
	if isFile(inTree) {
		candidates = append(candidates, inTree)
	}

	rootBin := filepath.Join(root, exe)
	if isFile(rootBin) {
		candidates = append(candidates, rootBin)
	}

	if len(candidates) == 0 {
		return ""
	}

	var best string
	var bestMtime time.Time
	for _, c := range candidates {
		fi, err := os.Stat(c)
		if err != nil {
			continue
		}
		if best == "" || fi.ModTime().After(bestMtime) {
			best = c
			bestMtime = fi.ModTime()
		}
	}
	return best
}

func guardEnabled(env map[string]string) bool {
	raw, ok := envLookup(env, "FLEET_DOGFOOD_GUARD")
	if !ok {
		return true
	}
	trimmed := strings.ToLower(strings.TrimSpace(raw))
	_, isOff := guardOffValues[trimmed]
	return !isOff
}

const pyDispatchRunner = `import importlib.util, os, sys, json
root = sys.argv[1]
p = os.path.join(root, 'tools', 'dispatch_worker.py')
try:
    spec = importlib.util.spec_from_file_location('dispatch_worker', p)
    if not (spec and spec.loader):
        raise ImportError(f"cannot load dispatch_worker from {p}")
    mod = importlib.util.module_from_spec(spec)
    sys.path.insert(0, os.path.dirname(p))
    spec.loader.exec_module(mod)
    raw = mod.build_command('probe', 'claude')
    e = json.loads(sys.argv[2])
    launch_cmd, guarded = mod.guarded_launch_command(raw, 'probe', 'claude', root, env=e)
    fak_bin = mod.resolve_fak_bin(root, e)
    live_on = mod.guard_enabled(e)
    print(json.dumps({'ok': True, 'guarded': bool(guarded), 'launch_cmd': list(launch_cmd), 'fak_bin': fak_bin or '', 'live_on': bool(live_on)}))
except (ImportError, AttributeError, OSError) as exc:
    print(json.dumps({'ok': False, 'moved': True, 'error': str(exc)}))
except Exception as exc:
    print(json.dumps({'ok': False, 'moved': False, 'error': str(exc)}))
`

func evalDispatchWorker(root string, env map[string]string) (guarded bool, launchCmd []string, fakBin string, liveOn bool, movedErr string) {
	dwPath := filepath.Join(root, "tools", "dispatch_worker.py")
	if _, err := os.Stat(dwPath); os.IsNotExist(err) {
		return false, nil, "", false, fmt.Sprintf("dispatch surface moved: cannot load dispatch_worker from %s", dwPath)
	}

	python, err := exec.LookPath("python3")
	if err != nil {
		python, err = exec.LookPath("python")
	}

	if err == nil {
		envBytes, _ := json.Marshal(env)
		cmd := exec.Command(python, "-c", pyDispatchRunner, root, string(envBytes))
		out, err := cmd.CombinedOutput()
		if err == nil {
			var resp struct {
				OK        bool     `json:"ok"`
				Moved     bool     `json:"moved"`
				Error     string   `json:"error"`
				Guarded   bool     `json:"guarded"`
				LaunchCmd []string `json:"launch_cmd"`
				FakBin    string   `json:"fak_bin"`
				LiveOn    bool     `json:"live_on"`
			}
			if jsonErr := json.Unmarshal(out, &resp); jsonErr == nil {
				if resp.OK {
					return resp.Guarded, resp.LaunchCmd, resp.FakBin, resp.LiveOn, ""
				}
				return false, nil, "", false, fmt.Sprintf("dispatch surface moved: %s", resp.Error)
			}
		}
	}

	contentBytes, err := os.ReadFile(dwPath)
	if err != nil {
		return false, nil, "", false, fmt.Sprintf("dispatch surface moved: %v", err)
	}
	content := string(contentBytes)
	for _, req := range []string{"guarded_launch_command", "build_command", "resolve_fak_bin", "guard_enabled"} {
		if !strings.Contains(content, req) {
			return false, nil, "", false, fmt.Sprintf("dispatch surface moved: module 'dispatch_worker' has no attribute '%s'", req)
		}
	}

	liveOn = guardEnabled(env)
	fakBin = resolveFakBin(root, env)
	guarded = liveOn && fakBin != ""
	if guarded {
		launchCmd = []string{fakBin, "guard", "--provider"}
	}
	return guarded, launchCmd, fakBin, liveOn, ""
}

// CountAuditRows counts decision rows the kernel recorded across the fleet's per-lane decision
// journals. Returns (rows, journals).
func CountAuditRows(root string, env map[string]string) (rows int, journals int) {
	var candidates []string
	fleetDir := filepath.Join(root, ".dispatch-runs", "guard-audit")
	if fi, err := os.Stat(fleetDir); err == nil && fi.IsDir() {
		entries, err := os.ReadDir(fleetDir)
		if err == nil {
			var jsonls []string
			for _, e := range entries {
				if !e.IsDir() && strings.HasSuffix(e.Name(), ".jsonl") {
					jsonls = append(jsonls, filepath.Join(fleetDir, e.Name()))
				}
			}
			sort.Strings(jsonls)
			candidates = append(candidates, jsonls...)
		}
	}

	if _, err := os.Stat(filepath.Join(root, ".git")); err == nil {
		cfg := envGet(env, "XDG_CONFIG_HOME")
		if cfg == "" {
			cfg = envGet(env, "APPDATA")
		}
		if cfg == "" {
			home := envGet(env, "HOME")
			if home == "" {
				if h, err := os.UserHomeDir(); err == nil {
					home = h
				}
			}
			if home != "" {
				cfg = filepath.Join(home, ".config")
			}
		}
		if cfg != "" {
			userJournal := filepath.Join(cfg, "fak", "guard-audit.jsonl")
			if isFile(userJournal) {
				candidates = append(candidates, userJournal)
			}
		}
	}

	for _, jp := range candidates {
		text, err := os.ReadFile(jp)
		if err != nil {
			continue
		}
		n := 0
		lines := strings.Split(string(text), "\n")
		for _, ln := range lines {
			if strings.TrimSpace(ln) != "" {
				n++
			}
		}
		if n > 0 {
			rows += n
			journals++
		}
	}
	return rows, journals
}

// DiagnoseAuditGap explains WHY the audit journal is empty. Returns "" when rows exist.
func DiagnoseAuditGap(root string) string {
	fleetDir := filepath.Join(root, ".dispatch-runs", "guard-audit")
	fi, err := os.Stat(fleetDir)
	if err != nil || !fi.IsDir() {
		return "no guard-audit journal directory yet — no guarded worker has run on this host (arm `fak guard -- <agent>` so the kernel records verdicts)"
	}
	entries, err := os.ReadDir(fleetDir)
	if err != nil {
		return "no guard-audit journal directory yet — no guarded worker has run on this host (arm `fak guard -- <agent>` so the kernel records verdicts)"
	}
	var jsonls []string
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".jsonl") {
			jsonls = append(jsonls, e.Name())
		}
	}
	sort.Strings(jsonls)
	if len(jsonls) == 0 {
		return "guard-audit directory exists but holds no journal files — the guard wire is configured but never exercised by a launched worker"
	}

	hasRows := false
	for _, j := range jsonls {
		text, err := os.ReadFile(filepath.Join(fleetDir, j))
		if err != nil {
			continue
		}
		for _, ln := range strings.Split(string(text), "\n") {
			if strings.TrimSpace(ln) != "" {
				hasRows = true
				break
			}
		}
		if hasRows {
			break
		}
	}
	if hasRows {
		return ""
	}

	return fmt.Sprintf("%d journal file(s) present but all blank — a guarded worker booted but proposed no adjudicated tool call (check the agent reached a tool use; an auth/login failure exits before the first verdict)", len(jsonls))
}

// Grade maps coverage percent and debt count onto an A-F letter grade.
func Grade(coverage float64, dogfoodDebt int) string {
	if dogfoodDebt == 0 && coverage >= 95.0 {
		return "A"
	}
	if dogfoodDebt == 0 {
		return "B"
	}
	if dogfoodDebt <= 1 {
		return "C"
	}
	if dogfoodDebt <= 2 {
		return "D"
	}
	return "F"
}

// FormatReport formats a Report into a human-readable string.
func FormatReport(r Report) string {
	lines := []string{
		fmt.Sprintf("dogfood-coverage: %.1f%% (%d/%d KPIs)  grade %s  dogfood_debt %d  audit_rows %d",
			r.Coverage, r.Met, r.Total, r.Grade, r.DogfoodDebt, r.AuditRows),
	}
	for _, k := range r.KPIs {
		mark := ".. "
		if k.OK {
			mark = "OK "
		} else if k.Hard {
			mark = "XX "
		}
		tag := "soft"
		if k.Hard {
			tag = "HARD"
		}
		lines = append(lines, fmt.Sprintf("  [%s] %-24s (%s)  %s", mark, k.Key, tag, k.Detail))
		if k.Evidence != "" {
			lines = append(lines, fmt.Sprintf("          -> %s", k.Evidence))
		}
	}
	if len(r.WorstFirst) > 0 {
		lines = append(lines, "  next: "+strings.Join(r.WorstFirst, ", "))
	}
	return strings.Join(lines, "\n")
}

// Evaluate computes the dogfood-coverage KPIs for workspace root under env.
func Evaluate(root string, env map[string]string) Report {
	if root == "" {
		root = findRepoRoot("")
	}

	var e map[string]string
	if env != nil {
		e = env
	} else {
		e = make(map[string]string)
		for _, kv := range os.Environ() {
			idx := strings.IndexByte(kv, '=')
			if idx >= 0 {
				e[kv[:idx]] = kv[idx+1:]
			}
		}
	}

	var kpis []KPI
	add := func(key string, ok, hard bool, detail, evidence string) {
		kpis = append(kpis, KPI{
			Key:      key,
			OK:       ok,
			Hard:     hard,
			Detail:   detail,
			Evidence: evidence,
		})
	}

	guarded, launchCmd, fakBin, liveOn, movedErr := evalDispatchWorker(root, e)
	if movedErr != "" {
		add("fleet_leaf_guarded", false, true,
			"dispatch_worker.guarded_launch_command fronts a claude worker with `fak guard`",
			movedErr)
		add("fak_bin_resolvable", false, true,
			"a `fak` binary resolves (FAK_BIN / tools/.bin / PATH) so guard-wrapping engages",
			movedErr)
		add("guard_default_on", false, true,
			"FLEET_DOGFOOD_GUARD is not disabled in the live environment (default ON)",
			movedErr)
	} else {
		leafEvidence := "UNWRAPPED (coverage=0 on this host)"
		if guarded {
			first := launchCmd
			if len(first) > 3 {
				first = first[:3]
			}
			leafEvidence = strings.Join(first, " ")
		}
		add("fleet_leaf_guarded", guarded, true,
			"dispatch_worker.guarded_launch_command fronts a claude worker with `fak guard`",
			leafEvidence)

		fakBinEvidence := fakBin
		if fakBinEvidence == "" {
			fakBinEvidence = "no fak binary found — workers run UNGUARDED"
		}
		add("fak_bin_resolvable", fakBin != "", true,
			"a `fak` binary resolves (FAK_BIN / tools/.bin / PATH) so guard-wrapping engages",
			fakBinEvidence)

		guardVal, hasGuardVal := e["FLEET_DOGFOOD_GUARD"]
		if !hasGuardVal {
			guardVal = "<unset=ON>"
		}
		add("guard_default_on", liveOn, true,
			"FLEET_DOGFOOD_GUARD is not disabled in the live environment (default ON)",
			"FLEET_DOGFOOD_GUARD="+guardVal)
	}

	wired := fileContains(filepath.Join(root, "tools", "issue_dispatch.py"), "guarded_launch_command")
	wiredEvidence := "MISSING"
	if wired {
		wiredEvidence = "tools/issue_dispatch.py calls guarded_launch_command"
	}
	add("issue_dispatch_wired", wired, true,
		"issue_dispatch.evaluate routes its detached spawn through guarded_launch_command",
		wiredEvidence)

	guardGo := isFile(filepath.Join(root, "cmd", "fak", "guard.go"))
	guardGoEvidence := "MISSING"
	if guardGo {
		guardGoEvidence = "cmd/fak/guard.go"
	}
	add("guard_verb_present", guardGo, true,
		"`fak guard -- claude` exists (cmd/fak/guard.go) as the one-command kernel front door",
		guardGoEvidence)

	documented := fileContains(filepath.Join(root, "DOGFOOD-CLAUDE.md"), "fak guard")
	add("guard_documented", documented, false,
		"DOGFOOD-CLAUDE.md documents the `fak guard` front door",
		"DOGFOOD-CLAUDE.md")

	rows, journals := CountAuditRows(root, e)
	auditEvidence := fmt.Sprintf("%d decision row(s) across %d journal(s)", rows, journals)
	if rows == 0 {
		auditEvidence += " — " + DiagnoseAuditGap(root)
	}
	add("audit_journal_evidence", rows > 0, false,
		"guarded workers have recorded kernel decisions in a durable audit journal",
		auditEvidence)

	doc := isFile(filepath.Join(root, "docs", "fak", "always-on-dogfood-server.md"))
	docEvidence := "MISSING"
	if doc {
		docEvidence = "docs/fak/always-on-dogfood-server.md"
	}
	add("always_on_server_doc", doc, false,
		"an always-on dogfood server design exists (Mac/GCP tiers)",
		docEvidence)

	plist := isFile(filepath.Join(root, "tools", "com.fak.serve-gateway.plist"))
	plistEvidence := "MISSING"
	if plist {
		plistEvidence = "tools/com.fak.serve-gateway.plist"
	}
	add("always_on_serve_plist", plist, false,
		"a launchd unit keeps a shared `fak serve` gateway alive 24/7",
		plistEvidence)

	total := len(kpis)
	met := 0
	var hardUnmet []KPI
	for _, k := range kpis {
		if k.OK {
			met++
		} else if k.Hard {
			hardUnmet = append(hardUnmet, k)
		}
	}
	coverage := 0.0
	if total > 0 {
		coverage = math.Round(1000.0*float64(met)/float64(total)) / 10.0
	}
	dogfoodDebt := len(hardUnmet)
	grade := Grade(coverage, dogfoodDebt)

	worstFirst := make([]string, 0, len(kpis))
	for _, k := range hardUnmet {
		worstFirst = append(worstFirst, k.Key)
	}
	for _, k := range kpis {
		if !k.Hard && !k.OK {
			worstFirst = append(worstFirst, k.Key)
		}
	}

	return Report{
		Schema:      Schema,
		Coverage:    coverage,
		Met:         met,
		Total:       total,
		DogfoodDebt: dogfoodDebt,
		Grade:       grade,
		AuditRows:   rows,
		KPIs:        kpis,
		WorstFirst:  worstFirst,
	}
}
